SignalServiceKit/src/Util/DataSource.m

```diff
@@ -0,0 +1,280 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+#import "DataSource.h"
+#import "MIMETypeUtil.h"
+
+NS_ASSUME_NONNULL_BEGIN
+
+@interface DataSourceValue ()
+
+@property (nonatomic) NSData *dataValue;
+
+@property (nonatomic) NSString *fileExtension;
+
+// This property is lazy-populated.
+@property (nonatomic) NSString *cachedFilePath;
+
+@property (nonatomic, nullable) NSString *sourceFilename;
+
+@end
+
+#pragma mark -
+
+@implementation DataSourceValue
+
++ (nullable id<DataSource>)dataSourceWithData:(NSData *)data fileExtension:(NSString *)fileExtension
+{
+    OWSAssert(data);
+
+    if (!data) {
+        return nil;
+    }
+
+    DataSourceValue *instance = [DataSourceValue new];
+    instance.dataValue = data;
+    instance.fileExtension = fileExtension;
+    return instance;
+}
+
++ (nullable id<DataSource>)dataSourceWithData:(NSData *)data utiType:(NSString *)utiType
+{
+    NSString *fileExtension = [MIMETypeUtil fileExtensionForUTIType:utiType];
+    return [self dataSourceWithData:data fileExtension:fileExtension];
+}
+
++ (nullable id<DataSource>)dataSourceWithOversizeText:(NSString *_Nullable)text
+{
+    OWSAssert(text);
+
+    if (!text) {
+        return nil;
+    }
+
+    NSData *data = [text dataUsingEncoding:NSUTF8StringEncoding];
+    return [self dataSourceWithData:data fileExtension:kOversizeTextAttachmentFileExtension];
+}
+
++ (id<DataSource>)dataSourceWithSyncMessage:(NSData *)data
+{
+    return [self dataSourceWithData:data fileExtension:kSyncMessageFileExtension];
+}
+
++ (id<DataSource>)emptyDataSource
+{
+    return [self dataSourceWithData:[NSData new] fileExtension:@"bin"];
+}
+
+- (NSData *)data
+{
+    OWSAssert(self.dataValue);
+
+    return self.dataValue;
+}
+
+- (nullable NSURL *)dataUrl
+{
+    NSString *_Nullable path = [self dataPath];
+    return (path ? [NSURL fileURLWithPath:path] : nil);
+}
+
+- (nullable NSString *)dataPath
+{
+    OWSAssert(self.dataValue);
+
+    @synchronized(self)
+    {
+        if (!self.cachedFilePath) {
+            NSString *dirPath = NSTemporaryDirectory();
+            NSString *fileName = [[[NSUUID UUID] UUIDString] stringByAppendingPathExtension:self.fileExtension];
+            NSString *filePath = [dirPath stringByAppendingPathComponent:fileName];
+            DDLogError(@"%@ ---- writing data", self.tag);
```

Comment:
This seems normal - why is this an error log?

Reply:
Removed.